File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int solve(TreeNode* root,int& ans){
        if(root==NULL)return 0;
        int left = solve(root->left,ans);
        int right = solve(root->right,ans);
        ans = max({ans, root->val, root->val+left+right});
        return max(0,root->val+max(left,right));
    }
    int maxPathSum(TreeNode* root) {
        int ans = -1e9;
        solve(root,ans);
        return ans;
    }
};

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int k;
    int solve(TreeNode root){
        if(root == null)return -1;
        int left = solve(root.left);
        k--;
        if(k == 0)return root.val;
        int right = solve(root.right);
        if(left != -1 )return left;
        if(right != -1 )return right;
        return -1;
    }
    public int kthSmallest(TreeNode root, int K) {
        k = K;
        return solve(root);
    }
}
